docs/proposals/optional_application_loging.md

```diff
@@ -0,0 +1,106 @@
+A log line doesn't contain OpenShift metadata, this is added to the message
+in the fluentd `k8_metadata_filter` plugin.
+
+#### Dynamic Control of Optional Logging
+
+By 'dynamic control' we understand the ability of an operator or a user to
+opt-in or opt-out certain OpenShift object from aggregated logging without
+having to redeploy fluentd.
+
+This approach has immediate effect, offers very granular level of control,
+is flexible, and doesn't bring any measurable latency for other containers logs.
+It requires additional resources and brings an overhead to the API server, where
+the fluentd input plugin has to `watch` certain API endpoints.
+
+Following schema describes how this can be implemented as an opt-out kind:
+- a user will label a namespace or a pod with `aggregated_logging_opt_out=true`
+- the fluentd will exclude log files based on this information
+![fluentd optin pipeline](https://raw.githubusercontent.com/wozniakjan/origin-aggregated-logging/proposal/optional_logging/docs/proposals/optin_pipeline.png)
+
+One way how to achieve this could be creating additional layer of symlinks to
+the log files. Currently the log files in `/var/log/containers/` are symlinks
+to files in `/var/log/pods/`, those are symlinks to files in `/var/lib/docker/containers/`,
+and finally those are UTF-8 formatted text files. We could create symlinks to
+`/var/log/containers/*` in a different directory, `/var/log/containers/optin/*`
+for instance. Then a daemon watching the OpenShift APIs and keep the symlinks
+in `/var/log/containers/optin/*` up to date.
+
+```
+monitored layer of symlinks                original layers of symlinks                UTF-8 text files
+           |                                   /                  \                          |
+           V                                  V                    V                         V
+/var/log/containers/optin/*  ->  /var/log/containers/*  ->  /var/log/pods/*  ->  /var/lib/docker/containers/*
+```
+
+Second approach vendoring custom changes in fluentd `in_tail` plugin is
+described in this implementation of a working prototype:
+- https://github.com/wozniakjan/origin-aggregated-logging/pull/2
+
+Fluentd `in_tail` plugin changes:
+- watch `/api/v1/namespaces` and `/api/v1/pods`
+- cache pod and namespace label `aggregated_logging_opt_out`
+- on notice, asynchronously call `refresh_watchers`
+  - otherwise would have to wait for a [`refresh_interval`](https://docs.fluentd.org/v0.12/articles/in_tail#refresh_interval)
+- in `expand_paths` filter out files matching cached notices
+
+#### Explicit Optional Logging
+
+Users and cluster operators would have to create an agreed precondition upon
+which logging will be gathered. Any change to the existing setup will require
+an action from a cluster operator and possibly redeployment of fluentd.
+
+This approach could be achieved using [`exclude_path`](https://docs.fluentd.org/v0.12/articles/in_tail#exclude_path)
```

Comment:
Are you saying this can be achieved now?  And it requires a config change and fluentd restart?

Reply:
Something along those lines. Given the `in_tail` plugin config is generated here:
https://github.com/openshift/origin-aggregated-logging/blob/829725d65b7139e79373b49d84c3333597287b3b/fluentd/generate_throttle_configs.rb#L100-L111

and is not included in the fluentd ConfigMap AFAIK, the process I tested is following:
- deploy logging
- modify `generate_throttle_configs.rb` to exclude desired pod with `exclude_path` in the `in_tail` plugin config
- rebuild and redeploy fluentd image

Not entirely trivial but trivial enough so I could do it. Also, I would imagine if we figure aout an elegant way of adding the the `in_tail` config to ConfigMap, that would remove the "rebuild the fluentd image" requirement.